notebooks/contrib-dev/Stochastic_Simu_Chem_Rxn.ipynb

```diff
@@ -0,0 +1,2654 @@
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## 3b-1.2 Crank-Nicolson (Trapezoid Rule)"
+      ],
+      "metadata": {
+        "id": "v9urNjZgaT1h"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Alternatively, the Trapezoid Rule is a numerical method used to approximate the solutions of ordinary differential equations (ODEs). It is a widely used method that provides a straightforward approach to solving ODEs when an exact analytical solution is not available.\n",
+        "\n",
+        "It also approximates the solution to a first-order ordinary differential equation of the form:\n",
+        "\n",
+        "\\begin{equation}\n",
+        "\\frac{dy}{dx} = f(x, y)\n",
+        "\\end{equation}\n",
+        "\n",
+        "The Trapezoid Rule also operates through an iterative approach that discretizes the continuous ODE into discrete steps by using the following formula to update the solution at each step:\n",
+        "\n",
+        "\\begin{equation}\n",
+        "y_{n+1} = y_n + \\frac{h}{2} \\left[f(x_n, y_n) + f(x_{n+1}, y_{n+1})\\right]\n",
+        "\\end{equation}\n",
+        "\n",
+        "In this formula:\n",
+        "\n",
+        "$y_{n+1}$ is the new approximation of the solution at the next step.\n",
+        "\n",
+        "$y_n$ is the previous approximation of the solution at the current step.\n",
+        "\n",
+        "$h$ is the step size, representing the size of the interval over which the solution is approximated.\n",
+        "\n",
+        "$x_n$ is the current value of the independent variable at the nth step.\n",
+        "\n",
+        "$f(x_n, y_n)$ and $f(x_{n+1}, y_{n+1})$ represent the rate of change of y at the current point $(x_n, y_n)$ and the predicted point $(x_{n+1}, y_{n+1})$, respectively."
+      ],
+      "metadata": {
+        "id": "r1EYn2umKwWW"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "**Now, implement the found differential equations into the Crank-Nicolson (Trapezoid Rule) to find and plot the  numerically solution.**"
+      ],
+      "metadata": {
+        "id": "ZqnzYbpfnvHm"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "import numpy as np\n",
+        "import matplotlib.pyplot as plt\n",
+        "\n",
+        "# Define the parameters and initial conditions\n",
+        "k1 = 0.5\n",
+        "k2 = 1.0\n",
+        "xa0 = 1000.0  # Initial concentration of Ca\n",
+        "t_max = 16.0  # Maximum time\n",
+        "dt = 0.1  # Time step\n",
+        "\n",
+        "# Define the rate equations\n",
+        "RHS_Ca = lambda Ca, Cb, t: -k1 * Ca\n",
+        "RHS_Cb = lambda Ca, Cb, t: k1 * Ca - k2 * Cb\n",
+        "RHS_Cc = lambda Cb, t: k2 * Cb\n",
+        "\n",
+        "# Function to perform Crank-Nicolson method\n",
+        "def crank_nicolson(RHS_Ca, RHS_Cb, RHS_Cc, Ca0, Cb0, Cc0, dt, numsteps, LOUD=False):\n",
```

Comment:
You should instead implement this as a generic function without the problem-specific information. You want to define your problem as returning a vector:
https://ndcbe.github.io/data-and-computing/notebooks/07/Trapezoid-Rule.html
https://ndcbe.github.io/data-and-computing/notebooks/07/Systems-of-Differential-Equations-and-Scipy.html#crank-nicolson

Reply:
This is a linear ODE. You should instead use our generic implementation of Crank-Nicolson from class:
https://ndcbe.github.io/data-and-computing/notebooks/07/Systems-of-Differential-Equations-and-Scipy.html#crank-nicolson